include/ensmallen_bits/nsgaiii/nsgaiii_impl.hpp

```diff
@@ -0,0 +1,320 @@
+	  {
+		newFitness.col(numMembers) = fitnessValues.col(fronts[l][i]);
+		nextPop.slice(numMembers++) = population.slice(fronts[l][i]);
+	  }
+	  l++;
+	}
+	size_t K = population - numMembers;
+	if (K == 0)
+	  continue;
+
+	// Normalization step.
+	arma::mat asf(function.NumObjectives(), function.NumObjectives(),
+		arma::fill::eye);
+
+	for (size_t i = 0; i < asf.n_rows; i++)
+	{
+	  for (size_t j = 0; j < asf.n_cols; j++)
+	  {
+		if (asf(i, j) == 0)
+		  asf(i, j) = 1e6;
+	  }
+	}
+
+	arma::mat f(function.NumObjectives(), fronts[0].size());
+	for (size_t i = 0; i < f.n_cols; i++)
+	  f.col(i) = fitnessValues.col(fronts[0][i]);
+
+	arma::vec zmin(function.NumObjectives());
+	for (size_t i = 0; i < function.NumObjectives(); i++)
+	  zmin[i] = f.row(i).min();
+
+  	for (size_t i = 0; i < fronts[0].size(); i++)
+	  f.col(i) -= zmin;
+
+	for (auto it = fitnessValues.begin(); it != fitnessValues.end(); it++)
+	  if (*it <= 1e-3)
+		*it = 0;
+
+	arma::mat max(asf.n_rows, fronts[0].size());
+	for (size_t i = 0; i < asf.n_rows; i++)
+	{
+	  arma::mat f_asf = f.t() * asf.row(i);
+	  for (size_t j = 0; j < f_asf.n_rows; j++)
+		max(i, j) = f_asf.row(j).max();
+	}
+
+	arma::vec mins(max.n_rows);
+	for (size_t i = 0; i < mins.n_elem; i++)
+	  mins[i] = max.row(i).index_max();
+
+	arma::mat extremePoints(function.NumObjectives(), function.NumObjectives());
+	for (size_t i = 0; i < function.NumObjectives(); i++)
+	  extremePoints.col(i) = f.col(mins[i]);
+
+	arma::mat b(extremePoints.n_rows, extremePoints.n_rows, arma::fill::ones);
+	arma::mat plane = arma::solve(extremePoints.t(), b);
+	arma::mat intercepts = 1 / plane;
+	fitnessValues.each_col() /= intercepts;
+
+	// Association step.
+	arma::mat perpDists(referenceSet.n_slices, numMembers);
+	arma::vec associates(numMembers);
+	for (size_t i = 0; i < numMembers; i++)
+	{
+	  double min = DBL_MAX;
+	  size_t minIdx = 0;
+	  for (size_t j = 0; j < referenceSet.n_slices; j++)
+	  {
+	  	arma::mat s = nextPop.slice(i);
+	  	arma::mat w = referenceSet.slice(j);
+	  	double wnorm = arma::norm(w, 2);
+	  	perpDists(j, i) = arma::norm(s - (w.t() * s * w) / std::pow(wnorm, 2), 2);
+	  	if (perpDists(j, i) < min)
+	  	{
+	  	  min = perpDists(j, i);
+	  	  minIdx = j;
+	  	}
+	  }
+	  associates[i] = minIdx;
+	}
+
+	// Compute niche count.
+	arma::uvec nicheCount(referenceSet.n_slices);
+	for (size_t i = 0; i < associates.n_elem; i++)
+	  nicheCount[i] = (arma::uword)arma::accu(arma::find(associates == i));
+
+	// Niching step.
+	size_t k = 1;
+	double min = nicheCount.min();
+	while (k <= K)
+	{
+	  std::vector<size_t> jmin;
+	  for (size_t i = 0; i < nicheCount.n_elem; i++)
+	  {
+	  	if (nicheCount[i] == min)
+	  	  jmin.push_back(i);
+	  }
+	  size_t randIdx = arma::randi<arma::uvec>(1, arma::distr_param(0, jmin.size()))[0];
+
+	  double distMin = DBL_MAX;
+	  size_t minIdx = 0;
+	  for (size_t i = 0; i < fronts[l].size(); i++)
+	  {
+		arma::mat s = nextPop.slice(i);
```

Comment:
We could avoid another copy if we directly use ``nextPop.slice(i)`` and ``referenceSet.slice(randIdx)``.